In [3]:
import numpy as np

# Step 1 : Read the Data!

In [17]:
#give path of the source-file and store that path in a variable named census
census = '/Users/muhammadahmad/Downloads/KS DS/makeSenseOfCensus.csv'


# Loading the data

You have studied getfromtext in Numpy Classes, now is the time to put it into action :) Below is the half completed code, you have to substitute values for delimiter and skip_header in it for hte data to be properly read into a Data array. The output of the array has been shared with you guys followed by the code block

In [19]:
#data = np.genfromtxt(data_file, delimiter = , skip_header = )
#print("\nData:\n\n ", data)


data=np.genfromtxt(census, delimiter=',', skip_header=1)

print("\nData:\n\n", data)


Data:

 [[39. 13.  4. ...  0. 40.  0.]
 [50. 13.  4. ...  0. 13.  0.]
 [38.  9.  4. ...  0. 40.  0.]
 ...
 [48. 13.  4. ...  0. 58.  1.]
 [40. 10.  4. ...  0. 40.  0.]
 [39. 13.  4. ...  0. 50.  1.]]


# Printing the type of data

In [21]:
#print the type of data
type(data)

numpy.ndarray

# Step 2 : Append the Data

A new record [[50,9,4,1,0,0,40,0]] has been given to you. You are supposed to append this new record into the data using np.concatenate()

In [ ]:
new_record = [[50,9,4,1,0,0,40,0]]

## Append 'new_record' (given) to 'data' using "np.concatenate()"

In [27]:
new_record = np.array([[50,9,4,1,0,0,40,0]])
data = np.concatenate((data, new_record), axis=0)

In [29]:
#display after appending the new record at the end of the array.
print("\nData after appending:\n\n", data)


Data after appending:

 [[39. 13.  4. ...  0. 40.  0.]
 [50. 13.  4. ...  0. 13.  0.]
 [38.  9.  4. ...  0. 40.  0.]
 ...
 [40. 10.  4. ...  0. 40.  0.]
 [39. 13.  4. ...  0. 50.  1.]
 [50.  9.  4. ...  0. 40.  0.]]


# Step 3 : Check if it's a young country or old country

 Create a new array called 'age' by taking only age column(age is the column with index 0) of 'census' array.

In [41]:
age=np.age = data[:, 0]
print("Age:",  age)

Age: [39. 50. 38. ... 40. 39. 50.]


Find the max age and store it in a variable called 'max_age'.



In [51]:
#print the max_age
max_age=np.max(age)
print("Max Age:", max_age)

Max Age: 90.0


Find the min age and store it in a variable called 'min_age'.

In [53]:
#print the min_age
min_age=np.min(age)
print("Min Age:", min_age)

Min Age: 17.0


Find the mean of the age and store it in a variable called 'age_mean'.

In [55]:
#print the age_mean
age_mean=np.mean(age)
print("Mean of Age:", age_mean)

Mean of Age: 38.06293706293706


Find the standard deviation of the age and store it in a variable called 'age_std'

In [57]:
#print the age_std
age_std=np.std(age)
print("Standard Deviation of Age:", age_std)

Standard Deviation of Age: 13.341478176165857


# Step 4: Let's check the country's race distribution to identify the minorities

Create four different arrays by subsetting 'census' array by Race column(Race is the column with index 2) and save them in 'race_0','race_1', 'race_2', 'race_3' and 'race_4' respectively(Meaning: Store the array where 'race'column has value 0 in 'race_0', so on and so forth)

In [59]:
race_0 = data[data[:, 2] == 0]
race_1 = data[data[:, 2] == 1]
race_2 = data[data[:, 2] == 2]
race_3 = data[data[:, 2] == 3]
race_4 = data[data[:, 2] == 4]

In [61]:
print("Race 0:\n", race_0)
print("Race 1:\n", race_1)
print("Race 2:\n", race_2)
print("Race 3:\n", race_3)
print("Race 4:\n", race_4)


Race 0:
 [[34.  4.  0.  1.  0.  0. 45.  0.]
 [35. 10.  0.  0.  0.  0. 40.  0.]
 [60.  6.  0.  0.  0.  0. 10.  0.]
 [51.  9.  0.  1.  0.  0. 40.  0.]
 [25. 10.  0.  0.  0.  0. 20.  0.]
 [20.  9.  0.  1.  0.  0. 48.  0.]
 [23. 10.  0.  0.  0.  0. 25.  0.]
 [29.  9.  0.  1.  0.  0. 40.  0.]
 [27. 12.  0.  1.  0.  0. 40.  0.]
 [31. 12.  0.  0.  0.  0. 40.  0.]]
Race 1:
 [[3.000e+01 1.300e+01 1.000e+00 1.000e+00 0.000e+00 0.000e+00 4.000e+01
  1.000e+00]
 [4.000e+01 1.100e+01 1.000e+00 1.000e+00 0.000e+00 0.000e+00 4.000e+01
  1.000e+00]
 [5.400e+01 1.000e+01 1.000e+00 1.000e+00 0.000e+00 0.000e+00 6.000e+01
  1.000e+00]
 [3.000e+01 9.000e+00 1.000e+00 0.000e+00 0.000e+00 1.573e+03 3.500e+01
  0.000e+00]
 [4.900e+01 1.000e+01 1.000e+00 1.000e+00 0.000e+00 0.000e+00 4.000e+01
  1.000e+00]
 [3.300e+01 1.300e+01 1.000e+00 0.000e+00 0.000e+00 0.000e+00 4.000e+01
  0.000e+00]
 [4.400e+01 1.400e+01 1.000e+00 0.000e+00 0.000e+00 0.000e+00 4.000e+01
  0.000e+00]
 [2.800e+01 9.000e+00 1.000e+00 0.00

Store the length of the above created arrays in 'len_0', 'len_1','len_2', 'len_3' and 'len_4' respectively

In [65]:
len_0=len(race_0)
len_0

10

In [71]:
len_1=len(race_1)
print(len_1)


27


In [73]:
len_2=len(race_2)
len_2



110

In [75]:
len_3=len(race_3)
len_3

6

In [77]:
len_4=len(race_4)
len_4

848

Find out which is the race with the minimum no. of citizens

In [129]:
lengths = [len_0, len_1, len_2, len_3, len_4]


min_length = np.min(lengths)
print("Race with Min no of Citizens",min_length)

Race with Min no of Citizens 6


Find out which is the race with the minimum no. of citizens

In [138]:
min_race_index = lengths.index(min_length)
print("Race with the minimum number of citizens (Race Index):", min_race_index)

Race with the minimum number of citizens (Race Index): 3


Store the number associated with the minority race in a variable called 'minority_race'(For eg: if "len(race_5)" is the minimum, store 5 in 'minority_race' because that is the index of the race having the least no. of citizens )

In [148]:
minority_race=len(race_3)
print(minority_race)

6


In [156]:
minority_race = lengths.index(min_length)
print("Race with the minimum number of citizens :", minority_race)

Race with the minimum number of citizens : 3


# Step 5: As per govt. records citizens above 60 should not work more than 25 hours a week. Let us check if the policy is in place

a.Create a new subset array called 'senior_citizens' by filtering 'census' according to age>60 (age is the column with index 0)

[Row,Columns]  :

In [160]:
#from age column, filter out age data > 60
senior_citizens = data[data[:, 0] > 60]
senior_citizens

array([[7.9000e+01, 1.0000e+01, 4.0000e+00, 1.0000e+00, 0.0000e+00,
        0.0000e+00, 2.0000e+01, 0.0000e+00],
       [6.7000e+01, 6.0000e+00, 4.0000e+00, 1.0000e+00, 0.0000e+00,
        0.0000e+00, 2.0000e+00, 0.0000e+00],
       [7.6000e+01, 1.4000e+01, 4.0000e+00, 1.0000e+00, 0.0000e+00,
        0.0000e+00, 4.0000e+01, 1.0000e+00],
       [6.1000e+01, 9.0000e+00, 4.0000e+00, 1.0000e+00, 0.0000e+00,
        0.0000e+00, 4.0000e+01, 0.0000e+00],
       [7.0000e+01, 1.0000e+01, 4.0000e+00, 1.0000e+00, 0.0000e+00,
        0.0000e+00, 4.0000e+01, 0.0000e+00],
       [6.4000e+01, 7.0000e+00, 4.0000e+00, 1.0000e+00, 0.0000e+00,
        2.1790e+03, 4.0000e+01, 0.0000e+00],
       [7.1000e+01, 1.0000e+01, 2.0000e+00, 1.0000e+00, 0.0000e+00,
        1.8160e+03, 2.0000e+00, 0.0000e+00],
       [6.8000e+01, 2.0000e+00, 4.0000e+00, 0.0000e+00, 0.0000e+00,
        0.0000e+00, 2.0000e+01, 0.0000e+00],
       [6.6000e+01, 1.1000e+01, 4.0000e+00, 0.0000e+00, 0.0000e+00,
        0.0000e+00, 2.0000e+

b.Add all the working hours(working hours is the column with index 6) of 'senior_citizens' and store it in a variable called 'working_hours_sum'

In [166]:
working_hours_sum = np.sum(senior_citizens[:, 6])
working_hours_sum

1917.0

c.Find the length of 'senior_citizens' and store it in a variable called 'senior_citizens_len'

In [170]:
senior_citizens_len=len(senior_citizens)
senior_citizens_len

61

d. Finally find the average working hours of the senior citizens by dividing 'working_hours_sum' by 'senior_citizens_len' and store it in a variable called 'avg_working hours'.

In [174]:
avg_working_hours=working_hours_sum/senior_citizens_len
avg_working_hours

31.42622950819672

e. Print 'avg_working_hours' and see if the govt. policy is followed.

In [176]:
print(avg_working_hours)

31.42622950819672


# Step 6: Let's check that higher educated people have better pay in general.

Create two new subset arrays called 'high' and 'low' by filtering 'census' according to education-num>10 and education-num<=10 (education-num is the column with index 1) respectively.



In [189]:
high = data[data[:, 1] > 10]
low = data[data[:, 1] <= 10]
print("High:\n", high)
print("Low:\n", low)

High:
 [[39. 13.  4. ...  0. 40.  0.]
 [50. 13.  4. ...  0. 13.  0.]
 [28. 13.  2. ...  0. 40.  0.]
 ...
 [45. 14.  4. ...  0. 45.  0.]
 [48. 13.  4. ...  0. 58.  1.]
 [39. 13.  4. ...  0. 50.  1.]]
Low:
 [[38.  9.  4. ...  0. 40.  0.]
 [53.  7.  2. ...  0. 40.  0.]
 [49.  5.  2. ...  0. 16.  0.]
 ...
 [56.  9.  4. ...  0. 55.  1.]
 [40. 10.  4. ...  0. 40.  0.]
 [50.  9.  4. ...  0. 40.  0.]]


Find the mean of income column(income is the column with index 7) of 'high' array and store it in 'avg_pay_high'. Do the same for 'low' array and store it's mean in 'avg_pay_low'.

In [197]:
avg_pay_high = np.mean(high[:, 7])  
avg_pay_low = np.mean(low[:, 7])    
print("Average income for high : ", avg_pay_high)
print("Average income for low :", avg_pay_low)

Average income for high :  0.42813455657492355
Average income for low : 0.13649851632047477


Average Pay for Higher Educated People 0.43

Average pay for low educated people 0.14